src/core/qgssnappingutils.cpp

```diff
@@ -92,8 +94,44 @@ bool QgsSnappingUtils::isIndexPrepared( QgsVectorLayer *vl, const QgsRectangle &
   return ( mStrategy == IndexHybrid || mStrategy == IndexExtent ) && loc->hasIndex() && ( !loc->extent() || loc->extent()->contains( aoi ) ); // the index - even if it exists - is not suitable
 }
 
+static bool _isMatchAVisibleLayer( const QgsPointLocator::Match &candidateMatch, QgsMapSettings mapSettings, bool snapOnInvisibleFeature )
+{
+
+  bool visible = true;
+  if ( !snapOnInvisibleFeature )
+  {
+    // segmentIntersection haven't a layer
+    if ( candidateMatch.layer() )
+    {
+      bool filter = false;
+      QgsRenderContext context( QgsRenderContext::fromMapSettings( mapSettings ) );
```

Comment:
Can you please move this up from _isMatchAVisibleLayer so that the render context and renderer is only constructed once per layer per snap? startRender is expensive so this is also a performance killer.